### ISSUE TITLE
Add a plugin interface to extend the editor

### ISSUE THREAD

**Issue by gracicot** [zed-industries/feedback#388]
### Check for existing issues

- [X] Completed

### Is your feature request related to a problem?

I would like to write a plugin to add CMake support. Currently, as far as I can see, there is no support for writing, browse or install plugins.

Plugins are an essential way to grow the ecosystem of developers and allow anyone to make the editor act as they need in way the Zed developers cannot afford to develop or haven't thought of.

### Describe the solution you'd like

I can imagine my plugin be able to do this:

 - Add commands to build the project, configure the build system and clear the build directory
 - Have the ability to instruct the language server where is the `compile_commands.json`
 - Use triggers to complete actions like updating the build system config when a build system related file is changed
 - Show compilation output and configuration output to the user
 - Provide a way to display some UI elements like selecting the build profile
 - Configure debuggers when Zed supports debugging
 - Running a given CMake executable (either from path or a config)

### If applicable, add mockups / screenshots to help present your vision of the feature

A good plugin system should be 100% multiplatform and also allow multiple languages. It should also be secure and sandboxed. Rust has an extremely good support for WebAssebly coupled with WASI. A plugin written in wasm would be fast, allow for plugins to be written in Rust and also be multiplatform so only one target is required for a plugin to be used everywhere. Wasm with wasmtime or wasmer has a controllable sandbox and it should be possible to expose an API for Rust, C, and Typescript to allow writing plugin in whatever the plugin developer likes to write.

**Comment by mikayla-maki** [comment]
Strongly agreed

**Comment by anthonyjclark** [comment]
> the capability for me to write a plug-in that would add my own take on Kakoune/Helix/etc. modal editing.

> layers but also sequences, combos, and all sorts of semi-intelligent features (e.g. a fun one we developed known as Caps Word)

> Being able to pair programmable keyboard features with programmable editor features is where I want to go.

I completely agree with these statements. I created my own modal setup for VS Code using https://github.com/DCsunset/vscode-modal-editor, and I'd love to do the same for Zed. I keep hoping for one of Zed, Lapce, or Fleet to add this functionality.

**Comment by mediabeastnz** [comment]
🤞

**Comment by jrmoynihan** [comment]
The easiest way to cut into VSCode's moat is to create an easy way to port those VSCode extensions into Zed.  An adaptor interface of some kind would go a long way.  A major advantage of a robust extension ecosystem is that it teaches the community what features users want, but don't have yet.  This would add additional value as a feedback loop for the Zed team.

**Comment by so1ve** [comment]
Maybe zed can load wasm plugins? like what dprint does?

**Comment by fkhoda** [comment]
[Extism](https://extism.org/) might be a good fit to support WASM-based plugins as well. It also includes support for multiple languages.

**Comment by avrame** [comment]
This is really the only issue holding me back from using Zed full time. With a plugin system, many of the other issues would be solved as the community could solve it by making it a plugin. I also think WASM is a good way to go so that the plugins can be written in pretty much any language.

**Comment by wzxu** [comment]
I think this is especially crucial to Zed because compared to VS Code, 1. Zed is closed source, and 2. not backed by a huge corporation. It's just impossible for Zed's developers alone to keep up the pace. Sublime Text's development is also slow and even went dead for a while, but with plugins it still remains pretty relevant.

**Comment by ayame113** [comment]
I think it might be possible to implement a JavaScript-based plugin system using Deno.

- [Roll your own JavaScript runtime - Deno](https://deno.com/blog/roll-your-own-javascript-runtime)
- [Roll your own JavaScript runtime, pt. 2 - Deno](https://deno.com/blog/roll-your-own-javascript-runtime-pt2)

**Comment by maddanio** [comment]
to also put a lower hanging froot into the room: a huge number of issues here would be solved by simply allowing new languages. that would not really need a plugin, just a way to detect them and specify an lsp
